### PR TITLE
Use 'go install' instead of 'go get' for installing binaries.

### PR DESCRIPTION
**Pull request by ybettan**
go get is deprecated for installing binaries

Also I have created a single target for installing all the binaries,
other targets won't install binaries themselves as this is not an
expected behavior from the user side.

Also removed the installation of
sigs.k8s.io/controller-runtime/tools/setup-envtest as it is not used
anyware.

Signed-off-by: Yoni Bettan <ybettan@redhat.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -60,11 +58,16 @@ all: $(SPECIALRESOURCE)
 
 ##@ Development
 
-manifests: controller-gen ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
-	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+go-install-binaries:
+	go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.6.1
+	go install golang.stackrox.io/kube-linter/cmd/kube-linter@v0.0.0-20210328011908-cb34f2cc447f
+	go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.43.0
```
Comment (ybettan) [inline Makefile]: You probably want that all of those to be in a separate script so you can call it from skipper as well I guess?
@yevgeny-shnaidman

Comment (yevgeny-shnaidman) [reply Makefile]: yes

Comment (ybettan) [reply Makefile]: Fixed.

File: Makefile.helper.mk [hunk 2]
```diff
@@ -46,16 +46,6 @@ e2e-test:
           KUBERNETES_CONFIG="$(KUBECONFIG)" go test -v -timeout 40m ./test/e2e/$$d -ginkgo.v -ginkgo.noColor -ginkgo.failFast || exit; \
         done
 
-# Download kube-linter locally if necessary
-KUBELINTER = $(shell pwd)/bin/kube-linter
```
Comment (qbarrand) [inline Makefile.helper.mk:50]: I think this comes straight from the `Makefile` that operator-sdk generates. Do we want to diverge from it again?

Comment (ybettan) [reply Makefile.helper.mk:50]: Can you elaborate on how it was generated and why we needed that generation?

**Comment by ybettan** [comment]
The failing GH actions are due to the fact that a PR changing the actions itself isn't tested on the actions in the PR but with the one that are already merged (this is by design due to security issue).

I tested those changes on [my fork's](https://github.com/ybettan/special-resource-operator/pull/3) CI (after the CI changes were made to that fork) to make sure the PR isn't breaking the CI.

The failures on the fork's CI are related to secrets for loging to docker and ocp and shouldn't be related to the PR itself.

File: Makefile.helper.mk (hunk 2 above)
Comment (qbarrand) [reply Makefile.helper.mk:50]: Sure!
I understand that the operator was created using the [Operator Framework](https://sdk.operatorframework.io/).
The framework includes `operator-sdk`, which [generates the initial Makefile](https://sdk.operatorframework.io/docs/building-operators/golang/quickstart/#steps).
When upgrading to a newer version of the framework, it is nice to have the Makefile as close to the original one as possible, as the [upgrade guide](https://sdk.operatorframework.io/docs/upgrading-sdk-version/) explains what changes should be made to it. Many changes to the Makefile are what made #116 quite painful.

Comment (ybettan) [reply Makefile.helper.mk:50]: Ummm last time I checked (long time ago) the operator-sdk wasn't the recommended tool for operators (I may be wrong and it may have changed).

1. How do we expect to use operator-sdk in the future?
2. I don't see any change log for using `go install` instead of the deprecated `go get` for binaries - can you point me to that direction? @qbarrand 

@filanov do we use `operator-sdk` in assisted?

Comment (qbarrand) [reply Makefile.helper.mk:50]: Is there another framework that we should use?

1. I can imagine that we will use some features at some point, and by using it we remain aligned with other operators that also use the framework.
2. Indeed, I ran `operator-sdk init` with the latest version and it still calls `go get`. But as long as it works, I would try not to touch the Makefile too much.

**Comment by ybettan** [comment]
/hold
Some of the makefile is generated by `operator-sdk` and we may need to get aligned to openshift standard if `operator-sdk` is the way to go with operators.

File: scripts/go_install_binaries.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+set -e
+
+go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.6.1
+go install golang.stackrox.io/kube-linter/cmd/kube-linter@v0.0.0-20210328011908-cb34f2cc447f
+go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.43.0
```
Comment (pmtk) [inline scripts/go_install_binaries.sh:7]: This installation method is not recommended:
https://golangci-lint.run/usage/install/#install-from-source
> Note: such go get installation aren't guaranteed to work. We recommend using binary installation

Maybe this is connected to the intermittent issues with `make lint` and should we use the recommended method of installation?

File: Makefile.helper.mk (hunk 2 above)
Comment (filanov) [reply Makefile.helper.mk:50]: This is what we install 
1. https://github.com/openshift/assisted-service/blob/fb2c51618d44bf6ed3855b1ad53c1ddbbcf1dc7c/hack/setup_env.sh#L57-L62
2. https://github.com/openshift/assisted-service/blob/fb2c51618d44bf6ed3855b1ad53c1ddbbcf1dc7c/hack/setup_env.sh#L70
not sure what is the operator used for, but to generate the CRDs we are using controller-gen https://github.com/openshift/assisted-service/blob/fb2c51618d44bf6ed3855b1ad53c1ddbbcf1dc7c/hack/generate.sh#L150

Comment (ybettan) [reply Makefile.helper.mk:50]: This is the operator sdk - https://github.com/openshift/assisted-service/blob/fb2c51618d44bf6ed3855b1ad53c1ddbbcf1dc7c/hack/setup_env.sh#L59 @filanov

Comment (ybettan) [reply Makefile.helper.mk:50]: @mhrivnak maybe you have some insights about it?

tl;dr:
1. Isn't it "unexpected" from the user side that using random makefile target, e.g. `make lint`, is installing binaries on the machine?
2. using `go get` for binaries is deprecated, does operator-sdk have a plane to replace those with `go install`?

Comment (mhrivnak) [reply Makefile.helper.mk:50]: operator-sdk is the recommended way to create operators. It extends [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/), which is responsible for scaffolding things like the [Makefile](https://github.com/kubernetes-sigs/kubebuilder/blob/57688c578b5d1189a7a1f28ebd1870ac9b534cd9/pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go).

One of the operator-sdk maintainers [made a PR to kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/pull/2383) converting to `go install`.

Comment (ybettan) [reply Makefile.helper.mk:50]: Ack. Thanks!

**Comment by ybettan** [comment]
After some [discussions](https://github.com/openshift-psap/special-resource-operator/pull/177#discussion_r762774898) seems that we should keep the makefile structure as it is.

Closing the PR for now.
